examples/gnn_link.py

```diff
@@ -0,0 +1,204 @@
+        data,
+        num_neighbors=num_neighbors,
+        time_attr="time",
+        input_nodes=(task.src_entity_table, src_node_indices),
+        input_time=torch.full(
+            size=(len(src_node_indices),), fill_value=seed_time, dtype=torch.long
+        ),
+        batch_size=2048,
+        shuffle=False,
+    )
+    dst_loader = NeighborLoader(
+        data,
+        num_neighbors=num_neighbors,
+        time_attr="time",
+        input_nodes=(task.dst_entity_table, torch.arange(task.num_dst_nodes)),
+        input_time=torch.full(
+            size=(task.num_dst_nodes,), fill_value=seed_time, dtype=torch.long
+        ),
+        batch_size=args.batch_size,
+        shuffle=False,
+    )
+    eval_loaders_dict[split] = (src_loader, dst_loader)
+
+model = Model(
+    data=data,
+    num_layers=args.num_layers,
+    channels=args.channels,
+    out_channels=args.channels,
+    aggr=args.aggr,
+    norm="layer_norm",
+).to(device)
+optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+
+
+def train() -> Dict[str, float]:
+    model.train()
+
+    loss_accum = count_accum = 0
+    for batch in tqdm(train_loader):
+        src_batch, batch_pos_dst, batch_neg_dst = batch
+        src_batch, batch_pos_dst, batch_neg_dst = (
+            src_batch.to(device),
+            batch_pos_dst.to(device),
+            batch_neg_dst.to(device),
+        )
+        x_src = model(src_batch, task.src_entity_table)
+        x_pos_dst = model(batch_pos_dst, task.dst_entity_table)
+        x_neg_dst = model(batch_neg_dst, task.dst_entity_table)
+
+        # [batch_size, ]
+        pos_score = torch.sum(x_src * x_pos_dst, dim=1)
+        if args.share_same_time:
+            # [batch_size, batch_size]
+            neg_score = x_src @ x_neg_dst.t()
+        else:
+            # [batch_size, ]
+            neg_score = torch.sum(x_src * x_neg_dst, dim=1)
+        diff_score = pos_score - neg_score
+        # BPR loss
+        optimizer.zero_grad()
+        loss = F.softplus(-diff_score).mean()
+        loss.backward()
+        optimizer.step()
+
+        loss_accum += loss.detach().item() * x_src.size(0)
+        count_accum += x_src.size(0)
+
+    return loss_accum / count_accum
+
+
+@torch.no_grad()
+def test(src_loader: NeighborLoader, dst_loader: NeighborLoader) -> np.ndarray:
+    model.eval()
+
+    dst_embs: list[Tensor] = []
+    for batch in tqdm(dst_loader):
+        batch = batch.to(device)
+        emb = model(batch, task.dst_entity_table).detach()
+        dst_embs.append(emb)
+    dst_emb = torch.cat(dst_embs, dim=0)
+    del dst_embs
+
+    mips = MIPSKNNIndex(dst_emb)
+
+    pred_index_mat_list: list[Tensor] = []
+    for batch in tqdm(src_loader):
+        batch = batch.to(device)
+        emb = model(batch, task.src_entity_table)
+        _, pred_index_mat = mips.search(emb, k=task.eval_k)
+        pred_index_mat_list.append(pred_index_mat.cpu())
+    pred = torch.cat(pred_index_mat_list, dim=0).numpy()
+    return pred
```

Comment:
would be great to support metric computation in batches, but I guess this is more of a nit.

Reply:
That's a good point. On backend, we are using pure `numpy` for metric computation (since we don't want to introduce `torchmetric` dependency). Can we leave the mini-batch computation as a follow-up when we actually hit memory issue?